quickwit/quickwit-search/src/root.rs

```diff
@@ -124,6 +130,103 @@ impl From<FetchDocsJob> for SplitIdAndFooterOffsets {
     }
 }
 
+/// Index metas needed for executing a leaf search request.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct IndexMetasForLeafSearch {
+    /// Index URI.
+    pub index_uri: Uri,
+    /// Doc mapper json string.
+    pub doc_mapper_str: String,
+}
+
+pub(crate) type IndexesMetasForLeafSearch = HashMap<IndexUid, IndexMetasForLeafSearch>;
+type TimestampFieldOpt = Option<String>;
+
+/// Validates request against each index's doc mapper and ensures that:
+/// - timestamp fields (if any) are equal across indexes.
+/// - resolved query ASTs are the same across indexes.
+/// Returns the timestamp field, the resolved query AST and the indexes metadatas
+/// needed for leaf search requests.
+/// Note: the requirements on timestamp fields and resolved query ASTs can be lifted
+/// but it adds complexity that does not seem needed right now.
+fn validate_request_and_build_metadatas(
+    indexes_metadatas: &[IndexMetadata],
+    search_request: &SearchRequest,
+) -> crate::Result<(TimestampFieldOpt, QueryAst, IndexesMetasForLeafSearch)> {
+    let mut metadatas_for_leaf: HashMap<IndexUid, IndexMetasForLeafSearch> = HashMap::new();
+    let query_ast: QueryAst = serde_json::from_str(&search_request.query_ast)
+        .map_err(|err| SearchError::InvalidQuery(err.to_string()))?;
+    let mut query_ast_resolved_opt: Option<QueryAst> = None;
+    let mut timestamp_field_opt: Option<String> = None;
+
+    for index_metadata in indexes_metadatas.iter() {
+        let doc_mapper = build_doc_mapper(
+            &index_metadata.index_config.doc_mapping,
+            &index_metadata.index_config.search_settings,
+        )
+        .map_err(|err| {
+            SearchError::InternalError(format!("Failed to build doc mapper. Cause: {err}"))
+        })?;
+        let query_ast_resolved_for_index = query_ast
+            .clone()
+            .parse_user_query(doc_mapper.default_search_fields())
+            // We convert the error to return a 400 to the user (and not a 500).
+            .map_err(|err| SearchError::InvalidQuery(err.to_string()))?;
+
+        // Validate uniqueness of resolved query AST.
```

Comment:
We had three possible specs here:
a) pick the `query_ast_resolved` of a given index.. Maybe the freshest one, or the first defined in the list.
b) use a different `query_ast_resolved` for each indexed.
c) return an error when facing an inconsistency (that's what you picked)

Obviously this is a trade-off, but I strongly prefer b).
I don't think there is a strong case to be defensive here.